File: executor/services/updater/binary_replacer.py
# SPDX-FileCopyrightText: 2025 Weibo, Inc.
#
# SPDX-License-Identifier: Apache-2.0

"""Binary replacer for executor self-update.

Downloads new binary and performs atomic replacement with backup.
"""

import logging
import os
import shutil
import stat
import tempfile
from pathlib import Path
from typing import Callable, Optional

from shared.utils.http_client import traced_session

logger = logging.getLogger(__name__)


class BinaryReplacer:
    """Download and replace executor binary with atomic operations.

    Downloads the new binary to a temporary location, creates a backup,
    and performs an atomic replacement to ensure the binary is never
    left in a corrupted state.
    """

    DOWNLOAD_TIMEOUT = 300  # 5 minutes for large binaries
    DOWNLOAD_CHUNK_SIZE = 8192  # 8KB chunks for streaming

    def __init__(self, download_url: str, auth_token: Optional[str] = None):
        """Initialize binary replacer.

        Args:
            download_url: URL to download the new binary from
            auth_token: Optional authentication token for the download API
        """
        self.download_url = download_url
        self.auth_token = auth_token

    def download_binary(
        self,
        progress_callback: Optional[Callable[[int, int], None]] = None,
    ) -> Path:
        """Download binary to temp location with streaming.

        Args:
            progress_callback: Optional callback(bytes_downloaded, total_bytes)
                Called periodically during download to report progress.

        Returns:
            Path to the downloaded temporary file

        Raises:
            RuntimeError: If download fails (network error, HTTP error, etc.)
        """
        headers = {}
        if self.auth_token:
            headers["PRIVATE-TOKEN"] = self.auth_token

        try:
            session = traced_session()
            response = session.get(
                self.download_url,
                headers=headers,
                timeout=self.DOWNLOAD_TIMEOUT,
                stream=True,
            )
            response.raise_for_status()

            # Get total size if available
            total_size = response.headers.get("content-length")
            if total_size:
                total_size = int(total_size)

            # Create temp file with restrictive permissions
            fd, temp_path = tempfile.mkstemp(suffix="-wegent-executor-new")
            os.close(fd)
            temp_file = Path(temp_path)

            try:
                downloaded = 0
                with open(temp_file, "wb") as f:
                    for chunk in response.iter_content(chunk_size=self.DOWNLOAD_CHUNK_SIZE):
                        if chunk:
                            f.write(chunk)
                            downloaded += len(chunk)
                            if progress_callback:
                                progress_callback(downloaded, total_size)

                return temp_file

            except Exception:
                # Clean up temp file on error
                if temp_file.exists():
                    temp_file.unlink()
                raise

        except Exception as e:
            raise RuntimeError(f"Failed to download binary: {e}")

    def replace_binary(self, new_binary: Path, current_binary: Path) -> bool:
        """Atomically replace current binary with backup.

        Performs the following steps:
        1. Create backup at {current_binary}.backup
        2. Set executable permissions on new binary (0o755)
        3. Atomically replace current binary with new one

        Args:
            new_binary: Path to the new binary (temp download location)
            current_binary: Path to the currently running binary

        Returns:
            True if replacement succeeded, False otherwise

        Note:
            os.replace() is atomic on POSIX systems. On Windows, it may
            fail if the binary is running - this is handled by the caller.
        """
        backup_path = current_binary.with_suffix(".backup")

        try:
            # Create backup of current binary if it exists
            if current_binary.exists():
                shutil.copy2(current_binary, backup_path)
                logger.debug(f"Created backup at {backup_path}")

            # Set executable permissions on new binary
            # 0o755 = rwxr-xr-x (owner can read/write/execute, group/others can read/execute)
            os.chmod(new_binary, stat.S_IRWXU | stat.S_IRGRP | stat.S_IXGRP | stat.S_IROTH | stat.S_IXOTH)

            # Atomic replace: new_binary -> current_binary
            # On POSIX, os.replace is atomic. On Windows, it may fail if binary is in use.
            os.replace(new_binary, current_binary)

            logger.info(f"Successfully replaced binary at {current_binary}")
            return True

        except PermissionError as e:
            logger.error(f"Permission denied during binary replacement: {e}")
            # Clean up on failure
            self._cleanup_on_failure(new_binary, backup_path, current_binary)
            return False
        except Exception as e:
            logger.error(f"Failed to replace binary: {e}")
            self._cleanup_on_failure(new_binary, backup_path, current_binary)
            return False

    def _cleanup_on_failure(
        self, new_binary: Path, backup_path: Path, current_binary: Path
    ) -> None:
        """Clean up temporary files on failure.

        Args:
            new_binary: Path to the new binary temp file
            backup_path: Path to the backup file
            current_binary: Path to the current binary
        """
        # Remove temp download if it exists
        if new_binary.exists():
            try:
                new_binary.unlink()
            except OSError:
                pass

        # If current binary was somehow removed, try to restore from backup
        if not current_binary.exists() and backup_path.exists():
            try:
                shutil.move(backup_path, current_binary)
                logger.warning("Restored binary from backup after failure")
            except OSError as e:
                logger.error(f"Failed to restore from backup: {e}")

    def cleanup_backup(self, current_binary: Path) -> bool:
        """Remove backup after successful update confirmation.

        Args:
            current_binary: Path to the current binary

        Returns:
            True if backup was removed or didn't exist, False on error
        """
        backup_path = current_binary.with_suffix(".backup")

        if not backup_path.exists():
            return True

        try:
            backup_path.unlink()
            logger.debug(f"Cleaned up backup at {backup_path}")
            return True
        except OSError as e:
            logger.warning(f"Failed to clean up backup: {e}")
            return False

    @staticmethod
    def format_progress_bar(downloaded: int, total: Optional[int], width: int = 40) -> str:
        """Format a progress bar for terminal display.

        Args:
            downloaded: Bytes downloaded so far
            total: Total bytes to download (None if unknown)
            width: Width of the progress bar in characters

        Returns:
            Formatted progress string like "[====>    ] 45% (23 MB / 50 MB)"
        """
        if total and total > 0:
            percent = min(100, int(100 * downloaded / total))
            filled = int(width * percent / 100)
            bar = "=" * filled + ">" + " " * (width - filled - 1)
            size_str = f"({downloaded // (1024 * 1024)} MB / {total // (1024 * 1024)} MB)"
            return f"[{bar}] {percent}% {size_str}"
        else:
            # Unknown total size
            mb = downloaded // (1024 * 1024)
            bar = "=" * width
            return f"[{bar}] {mb} MB downloaded"

File: executor/src/services/updater/binary_replacer.rs
// SPDX-FileCopyrightText: 2026 Weibo, Inc.
//
// SPDX-License-Identifier: Apache-2.0

use std::{
    fs::{self, File, OpenOptions},
    io::Write,
    path::{Path, PathBuf},
    time::{Duration, SystemTime, UNIX_EPOCH},
};

use futures_util::StreamExt;
use reqwest::header::{HeaderMap, HeaderValue};

const DOWNLOAD_TIMEOUT: Duration = Duration::from_secs(300);

#[derive(Debug, Clone)]
pub struct BinaryReplacer {
    download_url: String,
    auth_token: Option<String>,
}

impl BinaryReplacer {
    pub fn new(download_url: &str, auth_token: Option<&str>) -> Self {
        Self {
            download_url: download_url.to_owned(),
            auth_token: auth_token.map(ToOwned::to_owned),
        }
    }

    pub fn download_url(&self) -> &str {
        &self.download_url
    }

    pub fn auth_token(&self) -> Option<&str> {
        self.auth_token.as_deref()
    }

    pub async fn download_binary_to<F>(
        &self,
        target_dir: &Path,
        mut progress_callback: F,
    ) -> Result<PathBuf, BinaryReplaceError>
    where
        F: FnMut(u64, Option<u64>),
    {
        fs::create_dir_all(target_dir).map_err(BinaryReplaceError::download)?;
        let temp_path = self.temp_download_path(target_dir);
        let mut temp_file = OpenOptions::new()
            .write(true)
            .create_new(true)
            .open(&temp_path)
            .map_err(BinaryReplaceError::download)?;

        let result = self
            .stream_download(&mut temp_file, &mut progress_callback)
            .await;
        match result {
            Ok(()) => Ok(temp_path),
            Err(error) => {
                let _ = fs::remove_file(&temp_path);
                Err(error)
            }
        }
    }

    pub fn replace_binary(&self, new_binary: &Path, current_binary: &Path) -> bool {
        let backup_path = backup_path_for(current_binary);

        if let Some(parent) = current_binary.parent() {
            if fs::create_dir_all(parent).is_err() {
                return false;
            }
        }

        if current_binary.exists() && fs::copy(current_binary, &backup_path).is_err() {
            return false;
        }

        if set_executable_permissions(new_binary).is_err() {
            self.cleanup_on_failure(new_binary, &backup_path, current_binary);
            return false;
        }

        if fs::rename(new_binary, current_binary).is_err() {
            self.cleanup_on_failure(new_binary, &backup_path, current_binary);
            return false;
        }

        true
    }

    pub fn cleanup_backup(&self, current_binary: &Path) -> bool {
        let backup_path = backup_path_for(current_binary);
        if !backup_path.exists() {
            return true;
        }
        fs::remove_file(backup_path).is_ok()
    }

    pub fn format_progress_bar(downloaded: u64, total: Option<u64>, width: usize) -> String {
        match total.filter(|total| *total > 0) {
            Some(total) => {
                let percent = ((100 * downloaded) / total).min(100);
                let filled = ((width as u64 * percent) / 100) as usize;
                let arrow = if filled < width { ">" } else { "" };
                let padding = width.saturating_sub(filled + arrow.len());
                let bar = format!("{}{}{}", "=".repeat(filled), arrow, " ".repeat(padding));
                format!(
                    "[{bar}] {percent}% ({} MB / {} MB)",
                    downloaded / bytes_per_mb(),
                    total / bytes_per_mb()
                )
            }
            None => format!(
                "[{}] {} MB downloaded",
                "=".repeat(width),
                downloaded / bytes_per_mb()
            ),
        }
    }

    async fn stream_download<F>(
        &self,
        temp_file: &mut File,
        progress_callback: &mut F,
    ) -> Result<(), BinaryReplaceError>
    where
        F: FnMut(u64, Option<u64>),
    {
        let client = reqwest::Client::builder()
            .timeout(DOWNLOAD_TIMEOUT)
            .build()
            .map_err(BinaryReplaceError::download)?;
        let mut request = client.get(&self.download_url).headers(self.headers()?);
        request = request.timeout(DOWNLOAD_TIMEOUT);

        let response = request.send().await.map_err(BinaryReplaceError::download)?;
        let response = response
            .error_for_status()
            .map_err(BinaryReplaceError::download)?;
        let total = response.content_length();
        let mut downloaded = 0_u64;
        let mut stream = response.bytes_stream();

        while let Some(chunk) = stream.next().await {
            let chunk = chunk.map_err(BinaryReplaceError::download)?;
            if chunk.is_empty() {
                continue;
            }
            temp_file
                .write_all(&chunk)
                .map_err(BinaryReplaceError::download)?;
            downloaded += chunk.len() as u64;
            progress_callback(downloaded, total);
        }
        temp_file.flush().map_err(BinaryReplaceError::download)?;

        Ok(())
    }

    fn headers(&self) -> Result<HeaderMap, BinaryReplaceError> {
        let mut headers = HeaderMap::new();
        if let Some(token) = &self.auth_token {
            let value = HeaderValue::from_str(token).map_err(BinaryReplaceError::download)?;
            headers.insert("PRIVATE-TOKEN", value);
        }
        Ok(headers)
    }

    fn temp_download_path(&self, target_dir: &Path) -> PathBuf {
        let suffix = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|duration| duration.as_nanos())
            .unwrap_or_default();
        target_dir.join(format!(
            "{}-{suffix}-wegent-executor-new",
            std::process::id()
        ))
    }

    fn cleanup_on_failure(&self, new_binary: &Path, backup_path: &Path, current_binary: &Path) {
        let _ = fs::remove_file(new_binary);
        if !current_binary.exists() && backup_path.exists() {
            let _ = fs::rename(backup_path, current_binary);
        }
    }
}

#[derive(Debug, thiserror::Error)]
pub enum BinaryReplaceError {
    #[error("Failed to download binary: {0}")]
    Download(String),
}

impl BinaryReplaceError {
    fn download(error: impl ToString) -> Self {
        Self::Download(error.to_string())
    }
}

fn backup_path_for(current_binary: &Path) -> PathBuf {
    current_binary.with_extension("backup")
}

fn bytes_per_mb() -> u64 {
    1024 * 1024
}

#[cfg(unix)]
fn set_executable_permissions(path: &Path) -> std::io::Result<()> {
    use std::os::unix::fs::PermissionsExt;

    let mut permissions = fs::metadata(path)?.permissions();
    permissions.set_mode(0o755);
    fs::set_permissions(path, permissions)
}

#[cfg(not(unix))]
fn set_executable_permissions(_path: &Path) -> std::io::Result<()> {
    Ok(())
}
